README.md
# philosophy_blog

HISTORY.md

# Polymicrobial Systems Changelog

2018.04.21 / v0.16.0
====================

  * **[New]** Better Multi-Language Support!
    - contributed by **@Orhideous**
  * **[New]** Polymicrobial Systems can be served from non-root path (URI Prefix)
    - contributed by **@gugu**
  * **[Misc]** Upgrade to lunr v2.x
    - contributed by **@Orhideous**
  * **[Misc]** Code Refactor
    - contributed by **@Orhideous**
  * **[Misc]** Dependency upgrades

2018.03.29 / v0.15.0
====================

  * **[New]** Language Translations!
    - Romanian contributed by **@mariuspana**
  * **[Fixed]** #192 Any metadata will now cause metadata to render
    - **@mralexgray**
  * **[Fixed]** Login page loading of jQuery Backstretch plugin
    - **@Zezzty**
  * **[Fixed]** #247 Search result page no longer shows excerpt as link text
    - **@Zezzty**
  * **[Fixed]** #251 #194 Documentation in README for local install
    - **@shui**
  * **[Misc]** Dependency upgrades

2018.01.09 / v0.14.0
====================

  * **[New]** Language Translations!
    - Spanish contributed by **@dgarcia202**
    - Norwegian contributed by **@kek91**
    - Hungarian contributed by **@gabord**
  * **[New]** Multi-level Page Nesting
    - **@denisvmedia**
  * **[New]** Marking Active Category in UI
    - **@pmoleri**
  * **[New]** Export of Polymicrobial Systems class
    - **@pmoleri**
  * **[Improvement]** Search with Special Characters
    - **@cassiobsilva**
  * **[Improvement]** Upgrade to SweetAlert2
    - **@limonte**
  * **[Misc]** Remove Babel
    - **@pmoleri**
  * **[Misc]** Move from JSHint to ESLint
    - **@Sparticuz**
  * **[Misc]** Code Refinements
    - **@furier**
    - **@dettoni**
    - **@denisvmedia**
    - **@dgarcia202**
  * **[Misc]** Document Refinements
    - **@dgarcia202**
    - **@n7st**
  * **[Misc]** Dependency upgrades

2017.03.15 / v0.13.0
====================

  * **[New]** Nested Pages
    - contributed by **@zmateusz**
  * **[New]** Manual Category Title
    - contributed by **@theRealWardo**
  * **[New]** Last Edited Metadata Header
    - contributed by **@Sparticuz**
  * **[New]** Require Authentication for Viewing
    - contributed by **@bschne** and **@mohammadrafigh**
  * **[Improvement]** Meta Data RegEx Refinement
    - contributed by **@cmeyer90**
  * **[Improvement]** Unix Sitemap Generation
    - contributed by **@forsureitsme**
  * **[Improvement]** Display All Files Fix
    - contributed by **@forsureitsme**
  * **[Misc]** Code Refinements
    - **@shyim**
    - **@Sparticuz**
    - **@theRealWardo**
  * **[Misc]** Dependency upgrades

2016.09.13 / v0.11.0
====================

  * **[New]** Language Translations!
    - Mandarin Chinese contributed by **@noahvans**
    - French contributed by **@sfoubert**
    - Brazilian Portuguese contributed by **@ToasterBR**
  * **[New]** Google OAuth Support
    - contributed by **@Hitman666**
  * **[New]** Authentication for Edit (Public Read-Only)
    - contributed by **@alexspeller**
  * **[New]** Dynamic Sitemap.xml
    - contributed by **@sfoubert**
  * **[New]** Custom Variables
    - contributed by **@Sparticuz**
  * **[Improvement]** Multiple User Login
    - contributed by **@mohammadrafigh**
  * **[Improvement]** Table of Contents (Dynamic)
    - contributed by **@Sparticuz**
  * **[Misc]** Merged `Polymicrobial Systems-Core` module into repository
  * **[Misc]** Dependency upgrades


2016.06.18 / v0.10.1
====================

  * **[New]** Language Translations!
    - Right to Left support contributed by **@mohammadrafigh**
    - Persian contributed by **@mohammadrafigh**
  * **[New]** Docker support
    - contributed by **@prologic**
  * **[Improvement]** Better small-screen layout that automatically hides the left menu
    - contributed by **@ezaze**
  * **[Misc]** Upgrading raneto-core from v0.4.0 to v0.5.0


2016.05.22 / v0.10.0
====================

  * **[New]** Polymicrobial Systems Logo
    - contributed by **@mmamrila**
  * **[New]** Language Translations!
    - Russian contributed by **@iam-medvedev**
    - Turkish contributed by **@bleda**
  * **[New]** Metadata is editable
    - contributed by **@draptik**
  * **[Fixed] General BugFixes contributed by
    - **@draptik**
    - **@rogerhutchings**
    - **@dncrews**
    - **@durand**


2016.02.13 / v0.9.0
===================

  * **[Fixed]** Embedding images in content
    - contributed by **@helenamunoz**
  * **[Fixed]** Custom homepage via index.md file
    - contributed by **@dirivero**
  * **[Fixed]** Sanitizing file paths
  * **[New]** German Translation / Locale
    - contributed by **@Radiergummi**
  * **[New]** Authentication on Changes Only
    - contributed by **@Radiergummi**
  * **[New]** Vagrant Container
    - contributed by **@draptik**
  * **[New]** Category in Search Results
  * **[New]** Metadata on homepage
  * **[Upgraded]** Module raneto-core from v0.2.0 to v0.4.0
  * **[Upgraded]** Other Dependencies
  * **[Misc]** Broke up code into multiple files
  * **[Misc]** Delinted Code
  * **[Misc]** Overall refactor


2015.12.29 / v0.8.0
===================

  * **[Fixed]** URI Decoding with non-Latin characters
    - contributed by **@yaruson**
  * **[Fixed]** Windows compatability (use `npm run start_win`)
  * **[New]** Added Login Page to replace HTTP Basic Auth
    - contributed by **@matthiassb**
  * **[New]** Added ability to run Polymicrobial Systems as a PM2 service
    - contributed by **@matthiassb**
  * **[New]** Main Articles is now a category editable in the UI
    - contributed by **@yaruson**
  * **[New]** Using NPM for client-side libraries
    - contributed by **@sbussard**
  * **[Upgraded]** Improved Live Editor layout
    - contributed by **@draptik**
  * **[Removed]** Bower for client-side libraries
    - contributed by **@sbussard**

2015.10.11 / v0.7.1
===================

  * **[New]** Theme support. Copy `themes/default/` to `themes/<new name>/` and edit.
  * **[New]** Added toggle for enabling online editing of pages

test/functions.test.js
'use strict';

const fs = require('fs');
const path = require('path');
const chai = require('chai');
const expect = chai.expect;
const moment = require('moment');

const build_nested_pages = require('../app/functions/build_nested_pages.js');
const contentProcessors = require('../app/functions/contentProcessors');

const contentsHandler = require('../app/core/contents');
const utils = require('../app/core/utils');

chai.should();
chai.config.truncateThreshold = 0;

const config = {
  base_url: '',
  image_url: '/images',
  excerpt_length: 400,
  page_sort_meta: 'sort',
  category_sort: true,
  show_on_home_default: true,
  searchExtraLanguages: ['ru'],
  debug: false,
  content_dir: path.join(__dirname, 'content/'),
  datetime_format: 'Do MMM YYYY'
};

describe('#get_last_modified()', function () {

  it('returns last modified from page meta', async function () {
    const file_path = path.join(__dirname, 'content/page-with-bom-yaml.md');
    const content = fs.readFileSync(file_path, 'utf8');
    const modified = await utils.getLastModified(config, contentProcessors.processMeta(content), file_path);
    expect(modified).to.be.equal('14th Sep 2016');
  });

  it('returns last modified from fs', async function () {
    const file_path = path.join(__dirname, 'content/polymicrobial-systems-page.md');
    const content = fs.readFileSync(file_path, 'utf8');
    const modified = await utils.getLastModified(config, contentProcessors.processMeta(content), file_path);
    const fstime = moment(fs.lstatSync(file_path).mtime).format(config.datetime_format);
    expect(modified).to.be.equal(fstime);
  });

});

describe('#build_nested_pages()', function () {

  it('builds tree of pages', async function () {
    const pages = await contentsHandler(null, config);
    const result = build_nested_pages(pages);

    expect(result.length).to.be.equal(2);
    expect(result[1].files.length).to.be.equal(3);
    expect(result[1].files[0].files[0].title).to.be.equal('Sub2 Page');
  });

});


test/raneto-core.test.js

'use strict';

const path              = require('path');
const chai              = require('chai');
const expect            = chai.expect;
const contentProcessors = require('../app/functions/contentProcessors');

const searchHandler   = require('../app/core/search');
const pageHandler     = require('../app/core/page');
const contentsHandler = require('../app/core/contents');

chai.should();
chai.config.truncateThreshold = 0;

const config = {
  base_url: '',
  image_url: '/images',
  excerpt_length: 400,
  page_sort_meta: 'sort',
  category_sort: true,
  show_on_home_default: true,
  searchExtraLanguages: ['ru'],
  debug: false,
  content_dir: path.join(__dirname, 'content/'),
  datetime_format: 'Do MMM YYYY'
};

describe('#cleanString()', () => {

  it('converts "Hello World" into "hello-world"', () => {
    contentProcessors.cleanString('Hello World').should.equal('hello-world');
  });

  it('converts "/some/directory-polymicrobial-systems/hello/" into "some-directory-example-hello"', () => {
    contentProcessors.cleanString('/some/directory-polymicrobial-systems/hello/').should.equal('some-directory-example-hello');
  });

  it('converts "with trailing space " into "with-trailing-space"', () => {
    contentProcessors.cleanString('with trailing space ').should.equal('with-trailing-space');
  });

  it('converts "also does underscores" into "also_does_underscores"', () => {
    contentProcessors.cleanString('also does underscores', true).should.equal('also_does_underscores');
  });

  it('converts "/some/directory-polymicrobial-systems/underscores/" into "some_directory_example_underscores"', () => {
    contentProcessors.cleanString('/some/directory-polymicrobial-systems/underscores/', true).should.equal('some_directory_example_underscores');
  });

});

describe('#slugToTitle()', () => {

  it('converts "hello-world" into "Hello World"', () => {
    contentProcessors.slugToTitle('hello-world').should.equal('Hello World');
  });

  it('converts "dir/some-example-file.md" into "Some Example File"', () => {
    contentProcessors.slugToTitle('dir/some-example-file.md').should.equal('Some Example File');
  });

});

describe('#processMeta()', () => {

  it('returns array of meta values', () => {
    const result = contentProcessors.processMeta('/*\n' +
      'Title: This is a title\n' +
      'Description: This is a description\n' +
      'Sort: 4\n' +
      'Multi word: Value\n' +
      '*/\n');
    expect(result).to.have.property('title', 'This is a title');
    expect(result).to.have.property('description', 'This is a description');
    expect(result).to.have.property('sort', '4');
    expect(result).to.have.property('multi_word', 'Value');
  });

  it('returns an empty array if no meta specified', () => {
    const result = contentProcessors.processMeta('no meta here');
    /* eslint-disable no-unused-expressions */
    expect(result).to.be.empty;
  });

  it('returns proper meta from file starting with a BOM character', async () => {
    const result = await pageHandler(
      path.join(config.content_dir, 'page-with-bom.md'),
      config
    );
    expect(result).to.have.property('title', 'Example Page With BOM');
  });

  it('returns array of meta values (YAML)', () => {
    const result = contentProcessors.processMeta('---\n' +
      'Title: This is a title\n' +
      'Description: This is a description\n' +
      'Sort: 4\n' +
      'Multi word: Value\n' +
      '---\n');
    expect(result).to.have.property('title', 'This is a title');
    expect(result).to.have.property('description', 'This is a description');
    expect(result).to.have.property('sort', '4');
    expect(result).to.have.property('multi_word', 'Value');
  });

  it('returns proper meta from file starting with a BOM character (YAML)', async () => {
    const result = await pageHandler(
      path.join(config.content_dir, 'page-with-bom-yaml.md'),
      config
    );
    expect(result).to.have.property('title', 'Example Page With BOM for YAML');
  });

});

describe('#stripMeta()', () => {

  it('strips meta comment block', () => {
    const result = contentProcessors.stripMeta('/*\n' +
      'Title: This is a title\n' +
      'Description: This is a description\n' +
      'Sort: 4\n' +
      'Multi word: Value\n' +
      '*/\nThis is the content');
    result.should.equal('This is the content');
  });

  it('strips yaml meta comment block with horizontal rule in content', () => {
    const result = contentProcessors.stripMeta('---\n' +
      'Title: + This is a title\n' +
      '---\n' +
      'This is the content\n---');
    result.should.equal('This is the content\n---');
  });

  it('leaves content if no meta comment block', () => {
    const result = contentProcessors.stripMeta('This is the content');
    result.should.equal('This is the content');
  });

  it('leaves content with horizontal rule if no meta comment block', () => {
    const result = contentProcessors.stripMeta('This is the content\n---');
    result.should.equal('This is the content\n---');
  });

  it('only strips the first comment block', () => {
    const result = contentProcessors.stripMeta('/*\n' +
      'Title: This is a title\n' +
      'Description: This is a description\n' +
      'Sort: 4\n' +
      'Multi word: Value\n' +
      '*/\nThis is the content/*\n' +
      'Title: This is a title\n' +
      '*/');
    result.should.equal('This is the content/*\n' +
      'Title: This is a title\n' +
      '*/');
  });

});

describe('#processVars()', () => {

  it('replaces config vars in Markdown content', () => {
    const config = { base_url : '/base/url' };
    contentProcessors
      .processVars('This is some Markdown with a %base_url%.', config)
      .should.equal('This is some Markdown with a /base/url.');
  });

  it('replaces custom vars in Markdown content', () => {
    const config = {
      variables: [
        {
          name: 'test_variable',
          content: 'Test Variable'
        }
      ]
    };
    contentProcessors
      .processVars('This is some Markdown with a %test_variable%.', config)
      .should.equal('This is some Markdown with a Test Variable.');
  });

});

describe('#getPage()', () => {

  it('returns an array of values for a given page', async () => {
    const result = await pageHandler(
      path.join(config.content_dir, 'example-page.md'),
      config
    );
    expect(result).to.have.property('slug', 'example-page');
    expect(result).to.have.property('title', 'Example Page');
    expect(result).to.have.property('body');
    expect(result).to.have.property('excerpt');
  });

  it('returns null if no page found', async () => {
    const result = await pageHandler(
      path.join(config.content_dir, 'nonexistent-page.md'),
      config
    );
    /* eslint-disable no-unused-expressions */
    expect(result).to.be.null;
  });

});

describe('#getPages()', () => {

  it('returns an array of categories and pages', async () => {
    const result = await contentsHandler(null, config);
    expect(result[0]).to.have.property('is_index', true);
    expect(result[0].files[0]).to.have.property('title', 'Example Page');
    expect(result[1]).to.have.property('slug', 'sub');
    expect(result[1].files[0]).to.have.property('title', 'Example Sub Page');
  });

  it('marks activePageSlug as active', async () => {
    const result = await contentsHandler('/polymicrobial-systems-page', config);
    expect(result[0]).to.have.property('active', true);
    expect(result[0].files[0]).to.have.property('active', true);
    expect(result[1]).to.have.property('active', false);
    expect(result[1].files[0]).to.have.property('active', false);
  });

  it('adds show_on_home property to directory', async () => {
    const result = await contentsHandler(null, config);
    expect(result[0]).to.have.property('show_on_home', true);
  });

  it('adds show_on_home property to files', async () => {
    const result = await contentsHandler(null, config);
    expect(result[0].files[0]).to.have.property('show_on_home', true);
  });

  it('loads meta show_on_home value from directory', async () => {
    const result = await contentsHandler(null, config);
    expect(result[3]).to.have.property('show_on_home', false);
  });

  it('loads meta show_on_home value from file', async () => {
    const result = await contentsHandler(null, config);
    expect(result[0].files[4]).to.have.property('show_on_home', false);
  });

  it('applies show_on_home_default in absence of meta for directories', async () => {
    const result = await contentsHandler(null, Object.assign(config, {
      show_on_home_default: false
    }));
    expect(result[1]).to.have.property('show_on_home', false);
  });

  it('applies show_on_home_default in absence of meta for files', async () => {
    const result = await contentsHandler(null, Object.assign(config, {
      show_on_home_default: false
    }));
    expect(result[1].files[0]).to.have.property('show_on_home', false);
  });

  it('category index always shows on home', async () => {
    const result = await contentsHandler(null, Object.assign(config, {
      show_on_home_default: false
    }));
    expect(result[0]).to.have.property('show_on_home', true);
  });

});

describe('#doSearch()', () => {
  it('returns an array of search results', async () => {
    const result = await searchHandler('example', config);
    expect(result).to.have.length(5);
  });

  it('recognizes multiple languages', async () => {
    const result = await searchHandler('пример', config);
    expect(result).to.have.length(1);
  });

  it('returns an empty array if nothing found', async () => {
    const result = await searchHandler('qwerty', config);
    /* eslint-disable no-unused-expressions */
    expect(result).to.be.empty;
  });

});


app/index.js
'use strict';

// Modules
var path          = require('path');
var express       = require('express');
var favicon       = require('serve-favicon');
var logger        = require('morgan');
var cookie_parser = require('cookie-parser');
var body_parser   = require('body-parser');
var moment        = require('moment');
var hogan         = require('hogan-express');
var session       = require('cookie-session');
var passport      = require('passport');

function initialize (config) {

  // Load Translations
  if (!config.locale) { config.locale = 'en'; }
  if (!config.lang) {
    config.lang = require('./translations/' + config.locale + '.json');
  }

  // Content_Dir requires trailing slash
  if (config.content_dir[config.content_dir.length - 1] !== path.sep) { config.content_dir += path.sep; }

  // Load Files
  var authenticate              = require('./middleware/authenticate.js')               (config);
  var always_authenticate       = require('./middleware/always_authenticate.js')        (config);
  var authenticate_read_access  = require ('./middleware/authenticate_read_access.js')  (config);
  var error_handler             = require('./middleware/error_handler.js')              (config);
  var oauth2                    = require('./middleware/oauth2.js');
  var route_login               = require('./routes/login.route.js')                    (config);
  var route_login_page          = require('./routes/login_page.route.js')               (config);
  var route_logout              = require('./routes/logout.route.js')                   (config);
  var route_page_edit           = require('./routes/page.edit.route.js')                (config);
  var route_page_delete         = require('./routes/page.delete.route.js')              (config);
  var route_page_create         = require('./routes/page.create.route.js')              (config);
  var route_category_create     = require('./routes/category.create.route.js')          (config);
  var route_search              = require('./routes/search.route.js')                   (config);
  var route_home                = require('./routes/home.route.js')                     (config);
  var route_wildcard            = require('./routes/wildcard.route.js')                 (config);
  var route_sitemap             = require('./routes/sitemap.route.js')                  (config);

  // New Express App
  var app = express();
  var router = express.Router();

  // Set IP Address and Port
  app.set('host', process.env.HOST || '0.0.0.0');
  app.set('port', process.env.PORT || 5000);

  // set locale as date and time format
  moment.locale(config.locale);

  // Setup Views
  if (!config.theme_dir)  { config.theme_dir  = path.join(__dirname, '..', 'themes'); }
  if (!config.theme_name) { config.theme_name = 'default'; }
  if (!config.public_dir) { config.public_dir = path.join(config.theme_dir, config.theme_name, 'public'); }
  app.set('views', path.join(config.theme_dir, config.theme_name, 'templates'));
  app.set('layout', 'layout');
  app.set('view engine', 'html');
  app.enable('view cache');
  app.engine('html', hogan);

  // Setup Express
  app.use(favicon(config.public_dir + '/favicon.ico'));
  app.use(logger('dev'));
  app.use(body_parser.json());
  app.use(body_parser.urlencoded({ extended : false }));
  app.use(cookie_parser());
  app.use(express.static(config.public_dir));
  if (config.theme_dir !== path.join(__dirname, '..', 'themes')) {
    router.use(express.static(path.join(config.theme_dir, config.theme_name, 'public')));
  }
  router.use(config.image_url, express.static(path.normalize(config.content_dir + config.image_url)));
  router.use('/translations',  express.static(path.normalize(path.join(__dirname, 'translations'))));

  // HTTP Authentication
  if (config.authentication === true || config.authentication_for_edit || config.authentication_for_read) {
    app.use(session({
      secret            : config.secret,
      name              : 'raneto.sid',
      resave            : false,
      saveUninitialized : false
    }));
    app.use(authenticate_read_access);
    // OAuth2
    if (config.googleoauth === true) {
      app.use(passport.initialize());
      app.use(passport.session());
      router.use(oauth2.router(config));
      app.use(oauth2.template);
    }

    router.post('/rn-login', route_login);
    router.get('/logout', route_logout);
    router.get('/login',     route_login_page);
  }

  // Online Editor Routes
  if (config.allow_editing === true) {

    var middlewareToUse = authenticate;
    if (config.authentication_for_edit === true) {
      middlewareToUse = always_authenticate;
    }
    if (config.googleoauth === true) {
      middlewareToUse = oauth2.required;
    }

    router.post('/rn-edit',         middlewareToUse, route_page_edit);
    router.post('/rn-delete',       middlewareToUse, route_page_delete);
    router.post('/rn-add-page',     middlewareToUse, route_page_create);
    router.post('/rn-add-category', middlewareToUse, route_category_create);

  }

  // Router for / and /index with or without search parameter
  if (config.googleoauth === true) {
    router.get('/:var(index)?', oauth2.required, route_search, route_home);
    router.get(/^([^.]*)/, oauth2.required, route_wildcard);
  } else if (config.authentication_for_read === true) {
    router.get('/sitemap.xml', authenticate, route_sitemap);
    router.get('/:var(index)?', authenticate, route_search, route_home);
    router.get(/^([^.]*)/, authenticate, route_wildcard);
  } else {
    router.get('/sitemap.xml', route_sitemap);
    router.get('/:var(index)?', route_search, route_home);
    router.get(/^([^.]*)/, route_wildcard);
  }

  // Handle Errors
  router.use(error_handler);
  app.use(config.prefix_url || '/', router);

  // Wrap App if base_url is set
  if (config.base_url !== '' && config.nowrap !== true) {
    var wrap_app = express();
    wrap_app.set('port', process.env.PORT || 3000);
    wrap_app.use(config.base_url, app);
    return wrap_app;
  } else {
    return app;
  }

}

// Exports
module.exports = initialize;


app/routes/search.route.js

'use strict';

// Modules
var validator                      = require('validator');
var _s                             = require('underscore.string');
var remove_image_content_directory = require('../functions/remove_image_content_directory.js');

const searchHandler = require('../core/search');
const contentsHandler = require('../core/contents');

function route_search (config) {
  return async function (req, res, next) {

    // Skip if Search not present
    if (!req.query.search) { return next(); }

    // remove < and >
    var tagFreeQuery   = _s.stripTags(req.query.search);

    // remove /, ', " and & from query
    var invalidChars   = '&\'"/';
    var sanitizedQuery = validator.blacklist(tagFreeQuery, invalidChars);

    // trim and convert to string
    var searchQuery    = sanitizedQuery.toString(sanitizedQuery).trim();

    var searchResults  = await searchHandler(searchQuery, config);
    var pageListSearch = remove_image_content_directory(config, await contentsHandler(null, config));

    // TODO: Move to Polymicrobial Systems Core
    // Loop through Results and Extract Category
    searchResults.forEach(function (result) {
      result.category = null;
      var split = result.slug.split('/');
      if (split.length > 1) {
        result.category = split[0];
      }
    });

    return res.render('search', {
      config        : config,
      pages         : pageListSearch,
      search        : searchQuery,
      searchResults : searchResults,
      body_class    : 'page-search',
      lang          : config.lang,
      loggedIn      : ((config.authentication || config.authentication_for_edit) ? req.session.loggedIn : false),
      username      : ((config.authentication || config.authentication_for_edit) ? req.session.username : null)
    });

  };
}

// Exports
module.exports = route_search;


polymicrobial-systems/server.js
#!/usr/bin/env node

'use strict';

// Modules
var debug = require('debug')('raneto');

// Here is where we load Polymicrobial Systems.
// When you are in your own project repository,
// Polymicrobial Systems should be installed via NPM and loaded as:
// var raneto = require('raneto');
//
// For development purposes, we load it this way in this example:
var raneto = require('../app/index.js');

// Then, we load our configuration file
// This can be done inline, with a JSON file,
// or with a Node.js module as we do below.
var config = require('./config.default.js');

// Finally, we initialize Polymicrobial Systems
// with our configuration object
var app = raneto(config);

// Load the HTTP Server
var server = app.listen(app.get('port'), app.get('host'), function () {
  debug('Express HTTP server listening on port ' + server.address().port);
});


themes/default/public/scripts/login.js

jQuery(document).ready(function () {

  'use strict';

  var base_url = (typeof rn_base_url === "undefined") ? "" : rn_base_url;

  // Form validation
  $('.login-form input[type="text"], .login-form input[type="password"], .login-form textarea')
    .on('focus', function () {
      $(this).removeClass('input-error');
    });

  $('.login-form')
    .on('submit', function (e) {
      e.preventDefault();

      $(this).find('input[type="text"], input[type="password"], textarea').each(function () {
        if ($(this).val() === '') {
          e.preventDefault();
          $(this).addClass('input-error');
        } else {
          $(this).removeClass('input-error');
        }
      });

      if ($('.input-error').length === 0) {
        $.post(base_url + '/rn-login', $(this).serialize(), function (data) {

          Swal.fire({
            icon              : data.status ? 'success' : 'warning',
            title             : data.message,
            timer             : data.status ? 2000 : null,
            showConfirmButton : true
          });

          if (data.status) {
            window.setTimeout(function () {
              window.location = base_url + '/';
            }, 1500);
          }

        });
      }
    });

});


polymicrobial-systems/multiple-instances.js
#!/usr/bin/env node

'use strict';

// This example mounts 2 raneto instances with different configurations in the same server

// Modules
const debug = require('debug')('raneto');

// Here is where we load Polymicrobial Systems.
// When you are in your own project repository,
// Polymicrobial Systems should be installed via NPM and loaded as:
// var raneto = require('raneto');
//
// For development purposes, we load it this way in this example:
const raneto = require('../app/index.js');

// Load our base configuration file.
const config = require('./config.default.js');

const express = require('express');

// Create two subapps with different configurations
const appEn = raneto(Object.assign({}, config, { base_url : '/en', locale : 'en', nowrap : true }));
const appEs = raneto(Object.assign({}, config, { base_url : '/es', locale : 'es', nowrap : true }));

// Create the main app
const mainApp = express();
mainApp.use('/en', appEn);
mainApp.use('/es', appEs);

// Load the HTTP Server
const server = mainApp.listen(appEn.get('port'), appEn.get('host'), function () {
  debug('Express HTTP server listening on port ' + server.address().port);
});

// Now you can navigate to both:
// - http://localhost:3000/en
// - http://localhost:3000/es


polymicrobial-systems/config.default.js

'use strict';

// Modules
var path = require('path');

var config = {

  // Your site title (format: page_title - site_title)
  site_title: 'Polymicrobial Systems',

  // The base URL of your site (can use %base_url% in Markdown files)
  base_url: '',
  nowrap: true,

  // Used for the "Get in touch" page footer link
  support_email: '',

  // Footer Text / Copyright
  copyright: 'Copyright &copy; ' + new Date().getFullYear() + ' - <a href="http://polymicrobial.systems">Polymicrobial Systems</a>',

  // Excerpt length (used in search)
  excerpt_length: 400,

  // The meta value by which to sort pages (value should be an integer)
  // If this option is blank pages will be sorted alphabetically
  page_sort_meta: 'sort',

  // Should categories be sorted numerically (true) or alphabetically (false)
  // If true category folders need to contain a "sort" file with an integer value
  category_sort: true,

  // Controls behavior of home page if meta ShowOnHome is not present. If set to true
  // all categories or files that do not specify ShowOnHome meta property will be shown
  show_on_home_default: true,

  // Which Theme to Use?
  theme_dir  : path.join(__dirname, '..', 'themes'),
  theme_name : 'default',

  // Specify the path of your content folder where all your '.md' files are located
  // Fix: Needs trailing slash for now!
  // Fix: Cannot be an absolute path
  content_dir : path.join(__dirname, 'content'),

  // Where is the public directory or document root?
  public_dir  : path.join(__dirname, '..', 'themes', 'default', 'public'),

  // The base URL of your images folder,
  // Relative to config.public_dir
  // (can use %image_url% in Markdown files)
  image_url: '/images',

  // Add your analytics tracking code (including script tags)
  analytics: '',

  // Set to true to enable the web editor
  allow_editing : true,

  // Set to true to enable HTTP Basic Authentication
  authentication : true,

  // If editing is enabled, set this to true to only authenticate for editing, not for viewing
  authentication_for_edit: true,

  // If authentication is enabled, set this to true to enable authentication for reading too
  authentication_for_read: false,

  // Google OAuth
  googleoauth: false,
  google_group_restriction: {
    enabled: false,
    api_key: 'GOOGLE_API_KEY',
    group_name : 'GOOGLE_GROUP_NAME'
  },
  oauth2 : {
    client_id: 'GOOGLE_CLIENT_ID',
    client_secret: 'GOOGLE_CLIENT_SECRET',
    callback: 'http://localhost:3000/auth/google/callback',
    hostedDomain: 'google.com'
  },
  secret: 'someCoolSecretRightHere',

  credentials    : [
    {
      username : 'admin',
      password : 'password'
    },
    {
      username : 'admin2',
      password : 'password'
    }
  ],

  locale: 'en',

  // Support search with extra languages
  searchExtraLanguages: ['ru'],

  // Sets the format for datetime's
  datetime_format: 'Do MMM YYYY',

  // Set to true to render suitable layout for RTL languages
  rtl_layout: false,

  // Edit Home Page title, description, etc.
  home_meta : {
    // title       : 'Custom Home Title',
    // description : 'Custom Home Description'
  },

  // variables: [
  //   {
  //     name: 'test_variable',
  //     content: 'test variable'
  //   },
  //   {
  //     name: 'test_variable_2',
  //     content: 'test variable 2'
  //   }
  // ]

  // Set to true to enable generation of table of contents
  table_of_contents: false,

  // Configure generation of table of contents (see markdown-toc's docs for details on available options)
  table_of_contents_options: {
    // append: 'Table of contents appendix',
    // maxdepth: 6,
    // firsth1: true,
  },

  menu_on_pages: true,
  menu_on_page_collapsible: false
};

config.public_dir = path.join(__dirname, '..', 'themes', config.theme_name, 'public');

// Exports
module.exports = config;
